tests/admin.test.ts

```diff
@@ -0,0 +1,91 @@
+import { ActivityScope, ActivityType, UserAccessType } from '../types';
+import { ControllerFactory } from './controllers';
+import { DatabaseConnection, EventFactory, UserFactory, PortalState } from './data';
+
+beforeAll(async () => {
+  await DatabaseConnection.connect();
+});
+
+beforeEach(async () => {
+  await DatabaseConnection.clear();
+});
+
+afterAll(async () => {
+  await DatabaseConnection.clear();
+  await DatabaseConnection.close();
+});
+
+describe('retroactive attendance submission', () => {
+  test('logs activity, attendance, and points for users who have not attended', async () => {
+    const conn = await DatabaseConnection.get();
```

Comment:
Think you mentioned it earlier in the feedback testing PR but it slipped through, my bad. 

With this change the one thing that's fishy to me is having to await the factory fetch (e.g. `const adminController = await ControllerFactory.admin()`) for every time a controller is used, which doesn't make as much sense when reading it and which means you can't do something like `const response = await ControllerFactory.admin().submitAttendanceForUsers(...);`, since it needs 2 awaits (unless we use `.then` and `.catch` which breaks our codebase's conventions)